### ISSUE TITLE
if-let binding

### ISSUE THREAD

**Issue by kentosugama** [dfinity/motoko#3833]
In addition to `let-else` #3817, `if-let` would let users recover from pattern match failure.

Most likely implemented as direct desugaring in the parse to a `switch` with a wildcard case.

https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

**Comment by ggreif** [comment]
Since `if let ... { } else { }` is already valid syntax (but different semantics), we should consider a different keyword. Maybe
`case let <pat> = <exp> <then> else <else>` would fit the bill?

**Comment by rvanasa** [comment]
> Since `if let ... { } else { }` is already valid syntax (but different semantics), we should consider a different keyword. 

We should be okay to use `if let` because [parentheses are required around the let / let-else binding](https://embed.smartcontracts.org/motoko/g/JqqTGsAZ2Gh3uwh7ag7vps2bdsobMHHTtEzNy32Wy1g3Zax2UuHP?lines=2) with the current grammar (if I understand correctly).

**Comment by rossberg** [comment]
I really don't see how
```
if let <pat> = <exp1> { <exp2> } else { <exp3> }
```
is substantially shorter or clearer than the equivalent
```
switch <exp1> { case <pat> { <exp2> } case _ { <exp3> } }
```
It saves about one pair of braces. Does that justify piling up more redundant syntax that everybody needs to learn?

Don't be too quick to throw out new syntax unless it addresses a major pain point and reduces _structural_ complexity. Being small is a virtue for a language in itself. As is adhering to this principle from the [Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python#Principles):

    "There should be one – and preferably only one – obvious way to do it."

**Comment by nomeata** [comment]
For `if` (I assume) the `else` can be omitted? Just this morning I was writing some code (using `switch`) where I’d leave out the `else` if I had `if let`, but with `switch` I had to write a rather noisy and unhelpful `case _ ()`.

(But not feeling strongly about `iflet`.)

**Comment by rossberg** [comment]
Yeah, but that's still just minor convenience, probably still not worth the syntax budget.

If you truly want to go that route, you should at least come up with a syntax that scales, i.e., works naturally in similar other places, such as `for` or `while`. It's also preferable to have a stronger signal that this intentionally allows for partial patterns (unlike let). Something like `if (case pat = exp)` would hence be much better IMO, along with `for (case pat in exp)` and `while (case pat = exp)`.

**Comment by chenyan-dfinity** [comment]
Just my 2 cents. I don't feel strongly about having `if let` at the beginning. But since we already have `let else`, where the `else` can only be `None`, we kinda missing a more general sugar such as `if let`, which allows us to return a non-None value in the else branch.
